### PR TITLE
CLN/TYPE: window aggregation cleanups and typing

### PR DESCRIPTION
**Pull request by mroeschke**
- xref https://github.com/pandas-dev/pandas/pull/29878#issuecomment-561919909
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/_libs/window/aggregations.pyx [hunk 1]
```diff
@@ -794,14 +764,15 @@ def roll_kurt_fixed(ndarray[float64_t] values, ndarray[int64_t] start,
...
-                       bint is_monotonic_bounds=True):
+                       ndarray[int64_t] end, int64_t minp):
     cdef:
         float64_t val, prev
         float64_t x = 0, xx = 0, xxx = 0, xxxx = 0
         int64_t nobs = 0, i, j, s, e, N = len(values)
         ndarray[float64_t] output
+        bint is_monotonic_bounds
 
+    is_monotonic_bounds = np.all(np.diff(start) > 0) and np.all(np.diff(end) > 0)
```
Comment (WillAyd) [inline pandas/_libs/window/aggregations.pyx]: Can you make this a `cdef` function instead of copy?

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -412,7 +412,7 @@ def _get_roll_func(self, func_name: str) -> Callable:
             )
         return window_func
 
-    def _get_cython_func_type(self, func):
+    def _get_cython_func_type(self, func: str) -> Callable:
```
Comment (WillAyd) [inline pandas/core/window/rolling.py]: If possible to add subtypes to `Callable` typically very helpful

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_libs/window/aggregations.pyx [hunk 3]
```diff
@@ -183,14 +149,15 @@ cdef inline void remove_sum(float64_t val, int64_t *nobs, float64_t *sum_x) nogi
...
-                      bint is_monotonic_bounds=True):
+                      ndarray[int64_t] end, int64_t minp):
     cdef:
         float64_t sum_x = 0
         int64_t s, e
         int64_t nobs = 0, i, j, N = len(values)
         ndarray[float64_t] output
+        bint is_monotonic_bounds
 
+    is_monotonic_bounds = np.all(np.diff(start) > 0) and np.all(np.diff(end) > 0)
```
Comment (jbrockmendel) [inline pandas/_libs/window/aggregations.pyx]: is this call expensive?  we might have something in _libs.algos that short-circuits this check

File: pandas/_libs/window/aggregations.pyx [hunk 4]
```diff
@@ -504,7 +471,9 @@ def roll_var_variable(ndarray[float64_t] values, ndarray[int64_t] start,
         int64_t s, e
         Py_ssize_t i, j, N = len(values)
         ndarray[float64_t] output
+        bint is_monotonic_bounds
 
+    is_monotonic_bounds = np.all(np.diff(start) > 0) and np.all(np.diff(end) > 0)
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx]: make this a function

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/window/rolling.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/window/rolling.py]: Unfortunately the arguments to the cython functions for rolling are still not entirely uniform so the argument list can't be typed consistently.

File: pandas/_libs/window/aggregations.pyx (hunk 3 above)
Comment (mroeschke) [reply pandas/_libs/window/aggregations.pyx]: Good call. Was able to use `is_monotonic` in ` _libs.algos`

File: pandas/_libs/window/aggregations.pyx [hunk 5]
```diff
@@ -794,14 +766,15 @@ def roll_kurt_fixed(ndarray[float64_t] values, ndarray[int64_t] start,
...
-                       bint is_monotonic_bounds=True):
+                       ndarray[int64_t] end, int64_t minp):
     cdef:
         float64_t val, prev
         float64_t x = 0, xx = 0, xxx = 0, xxxx = 0
         int64_t nobs = 0, i, j, s, e, N = len(values)
         ndarray[float64_t] output
+        bint is_monotonic_bounds
 
+    is_monotonic_bounds = is_monotonic(start, False)[0] and is_monotonic(end, False)[0]
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx]: worth making this an inline helper function (not a big deal though)

**Comment by mroeschke** [comment]
All green.

**Review by WillAyd — APPROVED** [review]
lgtm

**Pull request merged**